### PR TITLE
Multi-monitor support

### LINKED ISSUE #10: Multi-Monitor Support
## Environment
- N/A

## Description of feature
I would like it to be possible for nody-greeter to open an electron BrowserWindow for each monitor / display that a user may have. I have a multi-monitor setup, and would like to create a theme that displays different information on each monitor, one of which will allow the user to log-in.

This would also allow for multi-monitor backgrounds to be displayed if themes could be configured to display different parts of a multi-monitor background on different displays. (I have plans for an animated SVG multi-monitor background to display behind my login screens, probably not too dissimilar to my website wallpaper here: https://samlanning.com/

## Possible solutions
I have already got a quick hacky solution working locally, but it wouldn't need too much cleanup before being PR-able I would say.

(I also have to say @JezerM, it was very quick to figure out how to start playing round with the code, thanks for keeping it clean and simple! I've wanted to play around with an electron and node based greeter for a while, and it's great to see it happening!)

It would roughly work by:

* creating an individual BrowserWindow for each display in `screen.getAllDisplays()`
* passing metadata to each browser window (probably via `_emit_signal`) that includes:
  * the monitor ID,
  * whether the monitor is primary
  * dimensions of the full set of moitors, and the position & size of the relevant monitor
* updating `_emit_signal` to send signals for every monitor
* (optional) allow for cross-window communication between BrowserWindows, for example if a theme allows changing a wallpaper in one window, it may need to notify other windows.

Then the default themes could be updated to only display the UI elements on the primary monitor, and display wallpaper only on other monitors.

If you're not opposed, I'd love to have a crack at this?

Cheers, and thanks again for the wonderful work!

### PR DESCRIPTION
**Pull request by s0**
closes #10

There are a couple of commits I also added to ease development a little bit (638c8524ebc6cc429968dd14faebf31034a410a8, 0d16e9dfa256c7b1dcc6d0fe42b5d685b666494d), I can more them to another PR or just remove if you'd prefer.

This is an initial implementation for multi-monitor support, that seems to be working quite well on my setup. I've included the ability for themes to get metadata about their window, including whether or not they're primary, and have hidden the login components for non-primary screens on the default themes.

I haven't yet implemented cross-window communication, but wanted to get this PR open as a draft before that to make sure that you're happy i'm on the right track. That's the last remaining part, and i'll use the dracula background-switcher as a demo of this functionality.

TODO:
* [x] Implement cross-window communication
* [x] Decide on mechanism for themes to pick per-window URLs
* [x] Fix sending metadata on theme refreshes

### CODE REVIEW

File: ts/preload.ts
Comment (JezerM) [inline ts/preload.ts:895]:
```suggestion
    nody_greeter: Nody | undefined;
```

The constants inside the Window interface should correspond to their existing variable. Thus, the `nody-greeter-types` package could suggest `nody_greeter` (which does exist), instead of `nody` (which doesn't).

File: ts/consts.ts
Comment (JezerM) [inline ts/consts.ts:8]:
```suggestion
    lightdm_signal: 'LightDMSignal',
```

It's just simpler.

File: ts/browser.ts
Comment (JezerM) [inline ts/browser.ts:135]:
```suggestion
  create_windows(): NodyWindow[] {
```

Change name to be consistent with its return type.

File: ts/browser.ts
Comment (JezerM) [inline ts/browser.ts:118]: Here, we could check if the window is in the main monitor to load `theme_url`, either load `secondary_theme_url`.

**Review by JezerM — COMMENTED** [review]
Overall, looks nice!

File: ts/preload.ts
Comment (s0) [reply ts/preload.ts:907]: ah whoops not sure how i missed this, think i was relying on TypeScript being in strict mode, and decided to rename the variable half way through! :sweat_smile: Nice catch!

File: ts/browser.ts
Comment (s0) [reply ts/browser.ts:115]: Interesting, we can definitely do this. How should we decide on what that secondary URL should be? I see a few options...

1. Only load a single URL (current state), and rely on theme checking metadata in javascript to decide what to display (what's currently done in 16f4bc2 for the existing themes)
2. Check for a `secondary.html` file, and load that if it exists, otherwise load `index.html`
3. Introduce a configuration file for themes as you suggest here: https://github.com/JezerM/nody-greeter/issues/10#issuecomment-1012668094

I don't have a preference either way, so am happy to implement what you suggest. If we go with option (3), i'd like a bit more guidance on how you'd want the config file to work, JSON/YAML, what filenames etc... any other configuration options...?

**Comment by s0** [comment]
Thanks for the initial review! Also noticed you introduced linting and formatting, very nice to see! I've incorporated your suggestions in a rebase, and also ensured that it's formatted + linted throughout the history.

I also spotted a bug where if a frame is reloaded (via ctrl+R, or via theme reloading etc...), then it won't initialize as metadata isn't sent more than once, so i'm going to look at fixing that now by having the theme frame request the metadata and sending it then, tracked this in a TODO above.

**Comment by s0** [comment]
Okay I've now updated the PR addressing the refresh bug, and allowing cross-window communication via a broadcast mechanism. This mechanism is then also used in the `dracula` theme to apply wallpaper changes across all screens.

Other than the open question around handling of a secondary URL, i think this is ready to go now :)

File: ts/browser.ts
Comment (JezerM) [reply ts/browser.ts:115]: I think I'll go with the third option, but others are also possible. I'm gonna push my changes later, but it would be in a YAML file, maybe `index.theme` (not sure yet), which could include the following options:

```yaml
# HTML file to use in main monitor
primary_html: "index.html"
# HTML file to use in non-main (secondary) monitors
# If not set or file does not exist, `primary_html` will be used
secondary_html: "secondary.html"
```

**Comment by JezerM** [comment]
Now, nody-greeter will look for `index.theme` (still not sure about the name) inside the theme's directory, if it does exist then it will be loaded and used to set `primary_url` and `secondary_url` for all monitors. However, if the `secondary_html` property does not exist, it will fallback to the `primary_html`.

**Comment by JezerM** [comment]
I just added the `secondary.html` and `index.theme` for both **gruvbox** and **dracula** themes.

Added a "fake" display to test it locally, and yeah, it works!

https://user-images.githubusercontent.com/59768785/149647671-5e6356b1-8284-4c5f-bd87-baa8b5109d4d.mov

File: ts/config.ts
Comment (s0) [inline ts/config.ts:179]: This doesn't accurately check whether the file paths end in `.html`, it will also match things like `foo_html` and `bar.html.something_else`. You'd need to use a regex that includes an end marker and escapes `.` like this: `/^.*\.html$/`. But you could also just use `.endsWith`:

```suggestion
    if (!theme_config.primary_html.endsWith(".html"))
      theme_config.primary_html = "index.html";
    if (!theme_config.secondary_html.endsWith(".html"))
      theme_config.secondary_html = undefined;
```

File: ts/config.ts
Comment (s0) [inline ts/config.ts:174]: Using `as` here means that we're making assumptions about the structure of a theme's / user's config file, which could lead to type errors later if they haven't done it correctly, it would be better to check that the structure of the value is of the type we expect.

There's a really cool library I've used in a number of production systems that allows you to define the types in combination with a runtime type-checker, called [`io-ts`](https://github.com/gcanti/io-ts/blob/master/index.md)

Then we could use it something like this:

```ts
import * as t from 'io-ts';

const THEME_CONFIG = t.type({
  /**
   * HTML file to use in main monitor
   * @example primary_html: "index.html"
   */
  primary_html: t.string;
  /**
   * HTML file to use in non-main (secondary) monitors
   * If the file does not exists or it's not set, `primary_html` will be used
   * @example secondary_html: "secondary.html"
   * @example secondary_html: ""
   */
  secondary_html: t.union([t.string, t.undefined]);
});

/// ...

const theme_config = yaml.load(file);
if (!THEME_CONFIG.is(theme_config)) {
  throw new Error('Invalid config').
}
```

We can also extend this functionality to automatically display an error message to the user indicating in what way their config is incorrect, using the built-in functionality of the library.

File: ts/config.ts
Comment (s0) [inline ts/config.ts:119]: I've just realised thata we're using synchronous / blocking filesystem operations throughout here. It would be better to use async functions, and use await [`fs.promises`](https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fs_promises_api), so that we don't freeze the Node.js process.

File: ts/config.ts
Comment (s0) [inline ts/config.ts:173]: it's bad practice to have a separate exists + read operation for a file, as the file could be deleted in between. Details: https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fs_exists_path_callback

Instead it's better to simply handle the error returned by the read operation.

File: ts/config.ts
Comment (s0) [inline ts/config.ts:202]: we could call load_theme_config() unconditionally, as it will need to handle the case where it can't read the file anyway.

**Review by s0 — COMMENTED** [review]
Cool, it will be useful functionality to have a theme config! :-D

> Now, nody-greeter will look for `index.theme` (still not sure about the name)

Hmm, i'm thinking it should probably end in `.yml` so that editors will at least pick up the correct syntax-highlighting for people, maybe `theme.yml` or `manifest.yml`?

File: ts/config.ts
Comment (s0) [reply ts/config.ts:174]: As a slight caveat to this, if we want `secondary_html` to be an optional parameter, we probably need to do something like this instead:

```ts
const THEME_CONFIG = t.intersection([
  t.type({
    /**
     * HTML file to use in main monitor
     * @example primary_html: "index.html"
     */
    primary_html: t.string,
  }),
  t.partial({
    /**
     * HTML file to use in non-main (secondary) monitors
     * If the file does not exists or it's not set, `primary_html` will be used
     * @example secondary_html: "secondary.html"
     * @example secondary_html: ""
     */
    secondary_html: t.string,
  })
]);
```

As a parameter existing with the value undefined is different to it not existing at all, and `.is()` will say it doesn't match...

File: ts/config.ts
Comment (JezerM) [reply ts/config.ts:60]: Yep, would be better.

File: ts/config.ts
Comment (JezerM) [reply ts/config.ts:179]: Sure, it wouldn't work xd

**Comment by JezerM** [comment]
`io-ts` looks interesting, powerful and easy to use. I did a little implementation with both greeter and theme's config, but I'd like you to check if this is correct or there's any improvement to be made.

File: ts/config.ts
Comment (s0) [inline ts/config.ts:23]: Any reason why you didn't do this instead?

```suggestion
    icon_theme: io_ts.union([io_ts.string, io_ts.undefined]),
    time_language: io_ts.union([io_ts.string, io_ts.undefined]),
```

If you have null, when these values don't exist it will probably fail the `decode()`

File: ts/config.ts
Comment (s0) [inline ts/config.ts:220]: Couple of things here to note regarding the decoding here...

* a value can fail `.is()` and succeed `.decode()` (for example decode can handle missing properties if that property is allowed to be undefined, because it returns its own value). As a result, `decode()` isn't guaranteed to produce errors that caused `.is()` to fail. So it's best to just simply call `.decode()` if we want to produce errors.
* you can just include the error message in the construction of the `Error()` rather than having a variable that you assign to sometimes.
* This code could be cleaned up a bit more if `validate_config` threw an error directly, or returned the valid value.

**Review by s0 — COMMENTED** [review]
Looking good, i think the code can be cleaned up a little more though. It's a bit hard to explain thoroughly with review comments so i'll push a commit in a bit that does what i'm thinking

File: ts/config.ts
Comment (s0) [reply ts/config.ts:22]: Have confirmed that this is the case, so `io_ts.undefined` is probably better here

**Comment by s0** [comment]
Just my one comment regarding `undefined` above probably needs to be addressed, then i think this is ready to go! :) Happy to defer the conversation around sync / async I/O operations to #13

File: ts/config.ts
Comment (JezerM) [reply ts/config.ts:22]: When `yaml` does the parse, it returns `null` instead of `undefined` on empty properties. So, it's better to handle null instead of undefined.

**Comment by JezerM** [comment]
Nice~ It looks better and simpler. Now, I'm going to check if there's any issue or anything to be improved before I merge this PR.

File: ts/config.ts
Comment (s0) [reply ts/config.ts:22]: Ah, in that case we should / could allow for both `null` and `undefined`, as if the line is missing from the config it will be undefined:

```suggestion
    icon_theme: io_ts.union([io_ts.string, io_ts.null, io_ts.undefined]),
    time_language: io_ts.union([io_ts.string, io_ts.null, io_ts.undefined]),
```

**Comment by s0** [comment]
> Nice~ It looks better and simpler. Now, I'm going to check if there's any issue or anything to be improved before I merge this PR.

Sweet! Excited to get this in! And thanks for being an easy maintainer to work with!

**Review by JezerM — APPROVED** [review]
I guess everything is alright!

Additionally, I just wanted to point out that if this feature is ever going to be added to **web-greeter**, a few things here could be changed (hopefully, just some nomenclature) to make them compatible~

**Pull request merged**
